### PR TITLE
Added secondary index support

### PR DESCRIPTION
**Pull request by sumocoder**
Class definition

```ruby
Class Post
  include Dynamoid::Document
  table name: :posts, key: :post_id, read_capacity: 200, write_capacity:  200
  range :posted_at, :datetime
  
  field :body, :string
  field :length, :string
  field :name, :string

  local_secondary_index :range_key => :name, :projected_attributes => :keys_only
  global_secondary_index :hash_key => :name, :range_key => :posted_at
  global_secondary_index :hash_key => :length, :read_capacity => 1, :write_capacity => 1
end

# Types of projected attributes
# - :keys_only
# - :all
# - [:body, :length, ...]
```

Query

The method will figure out which secondary index to use based on the parameters. If it cannot find an index, it will raise an exception.

```ruby
User.find_all_by_secondary_index(
  {:name => 'John'}, # hash
  :range => {"posted_at.lte" => Time.now.to_f} # range (optional)
)
```

Warning: The where clause is not yet supported. It needs more work for it to intelligently pick the right index.

### CODE REVIEW

File: dynamoid.gemspec [hunk 1]
```diff
@@ -71,5 +71,6 @@ Gem::Specification.new do |s|
   s.add_development_dependency(%q<github-markup>, [">= 0"])
   s.add_development_dependency(%q<pry>, [">= 0"])
   s.add_development_dependency(%q<coveralls>, [">= 0"])
+  s.add_development_dependency(%q<debugger>, [">= 0"])
```
Comment (philipmw) [inline dynamoid.gemspec]: This breaks Travis CI for any Ruby other than 1.9.x. Would you remove it?

Comment (sumocoder) [inline dynamoid.gemspec]: Removed the debugger

**Comment by beberry** [comment]
Exactly what I was looking for! Are there any updates on this? Is there a date when this could be expected to be released? Thanks!

**Comment by bglusman** [comment]
Likewise, :+1: , assumed this was already supported by Dynamoid and was just looking for documentation, hope this can be merged in soon or I may need to use this fork in my project, as a field that I planned as the primary key now looks like it needs to be an array of values, and I'll be doing lookups by these so definitely need an index...

**Comment by philipmw** [comment]
Thanks very much for this.

**Comment by bglusman** [comment]
@philipmw funny question, sorry to ask here, only tangentially related... is it possible to have assoiciations based on a key other than if in dynamoid at present?  or to have it know to use the primary key when the orimary key is NOT id?  I was struggling with this on Friday, and wasn't clear in the docs, your merge above just reminded me to clarify, as its related to why I may want/need dynamic secondary indexes.

**Comment by bglusman** [comment]
Oh, huh, it looks like this had probably legitimate test failures on several versions of ruby?

**Comment by bglusman** [comment]
Also, any target date for a new gem release that includes this merge?

**Comment by pboling** [comment]
Also hoping for a release with index support!

**Comment by bglusman** [comment]
@pboling you can use our dynamoid-edge gem if you like, we needed it sooner, so forked and released master (twice)

**Comment by pboling** [comment]
Oooh... I will certainly do that.  Thanks @bglusman!

**Comment by pboling** [comment]
@bglusman what is the fork?  The edge gem's homepage still lists this repo?

**Pull request merged**